File: solutions/503/main.py
from typing import List 

class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        result = [-1] * len(nums)
        st = [0]
        for i in range(1, len(nums)):
            while len(st) != 0 and nums[st[-1]] < nums[i]:
                result[st[-1]] = nums[i]
                st = st[:-1]
            st.append(i)
        for i in range(len(nums)):
            while len(st) != 0 and nums[st[-1]] < nums[i]:
                if result[st[-1]] == -1:
                    result[st[-1]] = nums[i]
                st = st[:-1]
        return result

if __name__ == "__main__": 
    nums = [1,2,3,4,3]
    ans = [2,3,4,-1,4]
    s = Solution()
    assert s.nextGreaterElements(nums) == ans 

File: solutions/503/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Sat Nov 27 16:53:29 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <algorithm> 
#include <stack> 
#include <gtest/gtest.h> 

using namespace std;

class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        vector<int> result(nums.size(), -1);
        stack<int> st;
        st.push(0);
        // simulate two rounds
        for (int i = 0; i < nums.size()*2; ++i) {
            if (nums[st.top()%nums.size()] < nums[i%nums.size()]) {
                while (!st.empty() && nums[st.top()%nums.size()] < nums[i%nums.size()]) {
                    result[st.top()%nums.size()] = nums[i%nums.size()];
                    st.pop();
                }
            }
            st.push(i);
        }
        return result;
    }
};

TEST(Test503, SimpleTest) {
    vector<int> nums{
        1,2,3,4,3
    };
    vector<int> ans{
        2,3,4,-1,4
    };
    Solution s; 
    vector<int> result = s.nextGreaterElements(nums);
    EXPECT_TRUE(equal(result.begin(), result.end(), ans.begin()));
}
